### PR TITLE
GitHub/108: Add missing run types 

### PR DESCRIPTION
**Pull request by jonasfh**
References issue #108

### CODE REVIEW

File: WebApp/WebContent/data_file/upload.xhtml [hunk 1]
```diff
@@ -10,10 +10,10 @@
 	<ui:define name="localHead">
 		<h:outputScript name="script/dataFiles.js"/>
 		<h:outputStylesheet name="style/dataFiles.css"/>
+		<h:outputStylesheet name="grid/grid.css" library="primefaces" />
 	</ui:define>
-
```
Comment (squaregoldfish) [inline WebApp/WebContent/data_file/upload.xhtml]: Note that the expression for getting the instrument name has changed. I think you need to rebase from upstream/alpha2.02 to get the updated code.

File: WebApp/src/uk/ac/exeter/QuinCe/web/files/UploadedDataFile.java [hunk 2]
```diff
@@ -134,6 +134,13 @@ public String getMessageClass() {
 		return messageClass;
 	}
 
+	public String getMissingRunTypeClass() {
+		if (null != dataFile && dataFile.getMissingRunTypes().size() > 0) {
+			return "show";
```
Comment (squaregoldfish) [inline WebApp/src/uk/ac/exeter/QuinCe/web/files/UploadedDataFile.java]: This should be `shown`

Comment (squaregoldfish) [reply WebApp/src/uk/ac/exeter/QuinCe/web/files/UploadedDataFile.java]: Incidentally, I think there's a few places where we reference CSS classes in the code. Should we have them statically defined somewhere? If you agree we can make a Medium Term issue for it.

File: WebApp/WebContent/data_file/upload.xhtml [hunk 3]
```diff
@@ -31,55 +31,116 @@
...
 							value="Errors"
 							>
-								<f:passThroughAttribute name="data-messages" value="#{file.messages}" />
+								<f:passThroughAttribute name="data-messages" value="#{uploadedFile.messages}" />
+							</p:commandButton>
+							<p:commandButton
+							icon="ui-icon-plus"
+							styleClass="missingRunTypeButton #{uploadedFile.missingRunTypeClass}"
+							onclick="PF('runTypesDialog_#{rowIndex}').show()"
+							title="Manage Missing Run Types"
```
Comment (squaregoldfish) [inline WebApp/WebContent/data_file/upload.xhtml]: Change to "Unrecognised Run Types Found"

File: WebApp/WebContent/data_file/upload.xhtml [hunk 4]
```diff
@@ -31,55 +31,116 @@
...
+							value="Manage Missing Run Types"
+							>
 							</p:commandButton>
+						    <p:dialog widgetVar="runTypesDialog_#{rowIndex}">
+								<div class="runTypeFile">
+									<p:fieldset legend="Update undefined Run Types">
```
Comment (squaregoldfish) [inline WebApp/WebContent/data_file/upload.xhtml]: I think the dialog needs a short sentence describing what the user's doing. Something like:

_This file contained unrecognised Run Types. Please assign them using the menu options._

There's a CSS class `instructions` that's used for these messages.

Then the `fieldset` title can change to "Unrecognised Run Types"

File: WebApp/WebContent/data_file/upload.xhtml [hunk 5]
```diff
@@ -31,55 +31,116 @@
...
+							</p:commandButton>
+							<p:commandButton
+							icon="ui-icon-plus"
+							styleClass="missingRunTypeButton #{uploadedFile.missingRunTypeClass}"
+							onclick="PF('runTypesDialog_#{rowIndex}').show()"
+							title="Manage Missing Run Types"
+							value="Manage Missing Run Types"
+							>
 							</p:commandButton>
+						    <p:dialog widgetVar="runTypesDialog_#{rowIndex}">
```
Comment (squaregoldfish) [inline WebApp/WebContent/data_file/upload.xhtml]: Make the dialog modal

**Comment by jonasfh** [comment]
- The missing x's are resolved in issue #743
- Improved user instructions as suggested
- Make sure button text is within bounds
-show -> shown
... all in commit 348b9eaef50634dd42764c9fadee49a1a0e9a2e4

**Comment by jonasfh** [comment]
There is a case sensitivity issue with the run rypes. The primary key, including the run_name column in the run_type table is case insensitive, so trying to insert ATM when atm exist, will fail. I am unsure if this is the desired behavior. If it is, values should be lowercased (or uppercased) on insertion, so the test for missing run types takes this into account.

**Comment by squaregoldfish** [comment]
My feeling is that run types should be case insensitive throughout. I can't think of a sensible reason for having two run types would differ only by case.

We'll have to make sure that this is handled at data extraction time too.

**Comment by jonasfh** [comment]
I think it is logical to handle the case sensitivity thing as a separate issue. It is not at all critical for this particular pull request, but should be handled. I will create another issue for this.

**Comment by jonasfh** [comment]
Added the case sensitivity fix from issue #748 to this pull request.

**Pull request merged**